File: Insertion_Sort/insert_sort.go
package insertionsort

type Pair struct {
	key   int
	value string
}

func insertionSort(pairs []Pair) [][]Pair {
	results := [][]Pair{}

	for i := 0; i < len(results); i++ {
		j := i - 1

		for j >= 0 && pairs[j+1].key < pairs[j].key {
			pairs[j+1], pairs[j] = pairs[j], pairs[j+1]
		}
		results = append(results, pairs)
	}

	return results
}

File: Insertion_Sort/insert_sort.js
class Pair {
	/**
	 * @param {number} key - The key to be stored in the pair
	 * @param {number} value - The value to be stored in the pair
	 */
	constructor(key, value) {
		this.key = key;
		this.value = value;
	}
}

class Solution {
	/**
	 * @param {Pair[]} pairs - The tuples of key, value pairs
	 * @returns {Pair[][]} returns the order of each sort process
	 */
	insertionSort(pairs) {
		const results = [];
		for (let i = 0; i < pairs.length; i++) {
			let j = i - 1;
			while (j >= 0 && pairs[j + 1].key < pairs[j].key) {
				[pairs[j], pairs[j + 1]] = [pairs[j + 1], pairs[j]];
				j--;
			}
			results.push([...pairs]);
		}

		return results;
	}
}
